File: php/103-binary-tree-zigzag-level-order-traversal.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution 
{
    /**
     * @param TreeNode $root
     * @return Integer[][]
     */
    public function zigzagLevelOrder($root) 
    {
        if (!$root) {
            return [];
        }
        $queue[] = [$root, 0];
        $data = [];
        while ($queue) {
            [$node, $level] = array_shift($queue);
            if (!isset($data[$level])) {
                $data[$level] = [];
            }
            if ($level % 2 == 0) {
                array_push($data[$level], $node->val);
            } else {
                array_unshift($data[$level], $node->val);
            }
            if ($node->left) {
                $queue[] = [$node->left, $level+1];
            }
            if ($node->right) {
                $queue[] = [$node->right, $level+1];
            } 
        } 
        return $data;
    }
}

File: go/103-binary-tree-zigzag-level-order-traversal.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type Node struct {
	tn    *TreeNode
	level int
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	resMap := make(map[int][]int)
	queue := []Node{Node{root, 0}}
	for len(queue) > 0 {
		n := queue[0]
		tn := n.tn
		level := n.level
		queue = queue[1:]
		if level%2 == 0 {
			resMap[level] = append(resMap[level], tn.Val)
		} else {
			resMap[level] = append([]int{tn.Val}, resMap[level]...)
		}
		if tn.Left != nil {
			queue = append(queue, Node{tn.Left, level + 1})
		}
		if tn.Right != nil {
			queue = append(queue, Node{tn.Right, level + 1})
		}
	}
	var res [][]int
	for i := 0; i < len(resMap); i++ {
		res = append(res, resMap[i])
	}
	return res
}
